### PR TITLE
Add the logic to set UID and Creation Time on the resources, add storage prefixes on policies, and store Calico metadata in the K8s annotations.

### PR DESCRIPTION
**Pull request by mgleung**
## Description
Added the handling for UID and CreationTimestamp to be added to the resources. Also added logic for prefixing the names internally on policies as well as logic for storing Calico metadata in the K8s annotations field.

## Todos
- [x] Tests
- [x] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

File: lib/backend/etcdv3/etcdv3.go [hunk 1]
```diff
@@ -85,6 +87,17 @@ func NewEtcdV3Client(config *apiconfig.EtcdConfig) (api.Client, error) {
 func (c *etcdV3Client) Create(ctx context.Context, d *model.KVPair) (*model.KVPair, error) {
 	logCxt := log.WithFields(log.Fields{"model-etcdKey": d.Key, "value": d.Value, "ttl": d.TTL, "rev": d.Revision})
 	logCxt.Debug("Processing Create request")
+
+	// Add in the UID and creation timestamp for the resource if needed since etcd does not do it itself.
```
Comment (robbrockbank) [inline lib/backend/etcdv3/etcdv3.go]: Not everything that gets here is a resource - so you can't cast to an ObjectMetaAccessor.

Any reason why you're not doing this processing in resources.go?

File: lib/backend/k8s/resources/bgppeer.go [hunk 2]
```diff
@@ -71,8 +71,7 @@ func (c BGPPeerConverter) ToKVPair(r Resource) (*model.KVPair, error) {
 
 	// Clear any CRD TypeMeta fields and then create a KVPair.
 	res := apiv2.NewBGPPeer()
-	res.ObjectMeta.Name = t.ObjectMeta.Name
-	res.ObjectMeta.Namespace = t.ObjectMeta.Namespace
+	res.ObjectMeta = t.ObjectMeta
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/bgppeer.go]: don't think this is correct we can't simply use the CRDs Metadata

**Comment by bcreane** [comment]
@mgleung - let's hold off merging this until after rc1 is cut.

File: lib/backend/k8s/resources/resources.go [hunk 3]
```diff
@@ -100,3 +102,74 @@ func SetCalicoMetadataFromK8sAnnotations(calicoRes Resource, k8sRes Resource) {
...
+	res.GetObjectMeta().SetFinalizers(meta.Finalizers)
+	res.GetObjectMeta().SetOwnerReferences(meta.OwnerReferences)
+	res.GetObjectMeta().SetClusterName(meta.ClusterName)
+
+	return nil
+}
+
+// Cleans out the Calico metadata from the annotations field. Needed to reduced side effects
+// that come from needing to modify the input data.
+func CleanAnnotations(res Resource) {
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/resources.go]: This doesn't need to be public.

It could also benefit from a better name - I wasn't sure what it was really doing at first.

Comment (robbrockbank) [reply lib/backend/k8s/resources/resources.go]: Maybe something like :   removeCalicoMetadataAnnotation

File: lib/backend/k8s/resources/resources.go [hunk 4]
```diff
@@ -100,3 +102,74 @@ func SetCalicoMetadataFromK8sAnnotations(calicoRes Resource, k8sRes Resource) {
 		}
 	}
 }
+
+// Store Calico Metadata in the in the k8s resource annotations for CRD backed resources.
+// This should store all Metadata except for those stored in Annotations and Labels and
+// store them in annotations.
+func StoreCalicoMetadataInK8sAnnotations(res Resource) error {
+	metadataBytes, err := json.Marshal(res.GetObjectMeta().(*metav1.ObjectMeta))
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/resources.go]: You don't need to cast this to a metav1.ObjectMeta do you?

File: lib/backend/k8s/resources/resources.go [hunk 5]
```diff
@@ -100,3 +102,74 @@ func SetCalicoMetadataFromK8sAnnotations(calicoRes Resource, k8sRes Resource) {
...
+	annotations[metadataAnnotation] = string(metadataBytes)
+	res.GetObjectMeta().SetAnnotations(annotations)
+
+	// Make sure to clear out all of the Calico Metadata out of the ObjectMeta except for
+	// Name, Namespace, and ResourceVersion.
+	res.GetObjectMeta().SetSelfLink("")
+	res.GetObjectMeta().SetUID("")
+	res.GetObjectMeta().SetGeneration(int64(0))
+	var emptyTime metav1.Time
+	res.GetObjectMeta().SetCreationTimestamp(emptyTime)
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/resources.go]: I don't think you need these var lines.  You can just do:  `res.GetObjectMeta().SetCreationTimestamp(metav1.Time{})`

same for the others below

File: lib/backend/k8s/resources/customresource.go [hunk 6]
```diff
@@ -60,6 +60,8 @@ func (c *customK8sResourceClient) Create(ctx context.Context, kvp *model.KVPair)
 
 	// Convert the KVPair to the K8s resource.
 	resIn, err := c.convertKVPairToResource(kvp)
+	// Clean out the annotations for the resource given since that's only used for our internal storage.
+	defer CleanAnnotations(resIn)
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/customresource.go]: I don't think this ClearAnnotations belongs here.  Everything that needs doing should be in the convertKVPairToResource.

File: lib/backend/k8s/resources/customresource.go [hunk 7]
```diff
@@ -99,6 +110,8 @@ func (c *customK8sResourceClient) Update(ctx context.Context, kvp *model.KVPair)
 	var updateError error
 	// Convert the KVPair to a K8s resource.
 	resIn, err := c.convertKVPairToResource(kvp)
+	// Clean out the annotations for the resource given since that's only used for our internal storage.
+	defer CleanAnnotations(resIn)
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/customresource.go]: Same here.

File: lib/backend/k8s/resources/customresource.go (hunk 6 above)
Comment (robbrockbank) [reply lib/backend/k8s/resources/customresource.go]: And in this case it feels particularly strange to use defer to do this since you are assuming that resIn is what we return below - which makes things hard to maintain.  This would be better handled just before we return the result.

In any case - the convertResourceToKVPair should take care of the converting back to the Calico ObjectMeta.

File: lib/backend/k8s/resources/resources.go [hunk 8]
```diff
@@ -100,3 +102,74 @@ func SetCalicoMetadataFromK8sAnnotations(calicoRes Resource, k8sRes Resource) {
...
+	annotations := res.GetObjectMeta().GetAnnotations()
+	if len(annotations) == 0 {
+		annotations = make(map[string]string)
+	}
+	annotations[metadataAnnotation] = string(metadataBytes)
+	res.GetObjectMeta().SetAnnotations(annotations)
+
+	// Make sure to clear out all of the Calico Metadata out of the ObjectMeta except for
+	// Name, Namespace, and ResourceVersion.
+	res.GetObjectMeta().SetSelfLink("")
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/resources.go]: this would have been less code if you'd first got the ObjectMeta in a local variable (and potentially less operations depending on how clever the compiler is)

File: lib/backend/k8s/resources/resources.go (hunk 3 above)
Comment (robbrockbank) [reply lib/backend/k8s/resources/resources.go]: IN fact I think I would just scrap this method and build it into the method above - that way those two methods are more symmetric.

One updates Metadata for K8s
One updates Metadata for Calico

Comment (robbrockbank) [reply lib/backend/k8s/resources/resources.go]: With that in mind, perhaps naming those methods something like:
```
ConvertMetadataCalicoToK8s()
ConvertMetadataK8sToCalico()
```

File: lib/backend/k8s/resources/resources.go [hunk 9]
```diff
@@ -100,3 +102,74 @@ func SetCalicoMetadataFromK8sAnnotations(calicoRes Resource, k8sRes Resource) {
...
+		return errors.New("Attempting to create a resource from a resource without Calico metadata set in the annotations.")
+	}
+	meta := &metav1.ObjectMeta{}
+	err := json.Unmarshal([]byte(annotations[metadataAnnotation]), meta)
+	if err != nil {
+		return err
+	}
+
+	// Overwrite the K8s metadata with the Calico metadata.
+	res.GetObjectMeta().SetSelfLink(meta.SelfLink)
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/resources.go]: we never store a self link

File: lib/backend/k8s/resources/resources.go [hunk 10]
```diff
@@ -100,3 +102,74 @@ func SetCalicoMetadataFromK8sAnnotations(calicoRes Resource, k8sRes Resource) {
...
+	var emptyTime metav1.Time
+	res.GetObjectMeta().SetCreationTimestamp(emptyTime)
+	res.GetObjectMeta().SetDeletionTimestamp(nil)
+	res.GetObjectMeta().SetDeletionGracePeriodSeconds(nil)
+	res.GetObjectMeta().SetInitializers(nil)
+	var emptyStringSlice []string
+	res.GetObjectMeta().SetFinalizers(emptyStringSlice)
+	var emptyReferences []metav1.OwnerReference
+	res.GetObjectMeta().SetOwnerReferences(emptyReferences)
+	res.GetObjectMeta().SetClusterName("")
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/resources.go]: All of this field copying fields like a maintenance hassle - especially if k8s decide to add more fields to the Metadata.

I think a better a more maintainable way to do this would be to use the DeepCopyInto() method defined on the ObjectMeta struct.  You could create your own local "empty" ObjectMeta with just the fields that should be included:  ie. the name, namespace, labels and annotations - and then use DeepCopyInto to copy over the Metadata in the resource.  This is more resilient to new fields being added.

Comment (robbrockbank) [reply lib/backend/k8s/resources/resources.go]: Same goes for both of these conversion functions.

Comment (robbrockbank) [reply lib/backend/k8s/resources/resources.go]: I think you should also avoid storing duplicate copies of everything - so I think the ObjectMeta that is stored in the annotation should have the Name, Namespace, TypeMeta, Annotations and Labels removed before marshaling into json.

Obviously you'll need to add those fields back in the other function.

File: lib/clientv2/globalnetworkpolicy.go [hunk 11]
```diff
@@ -113,3 +147,21 @@ func defaultPolicyTypesField(spec *apiv2.PolicySpec) {
...
+func removePrefixFromGlobalPolicy(policy *apiv2.GlobalNetworkPolicy) {
+	name := policy.GetObjectMeta().GetName()
+	if !strings.HasPrefix(name, "knp.") {
+		parts := strings.SplitN(name, ".", 2)
+		resetGlobalPolicyName(policy, parts[1])
+	}
+}
+
+func resetGlobalPolicyName(res *apiv2.GlobalNetworkPolicy, originalName string) {
+	res.GetObjectMeta().SetName(originalName)
```
Comment (robbrockbank) [inline lib/clientv2/globalnetworkpolicy.go]: Not sure what the point of this function is.

File: lib/backend/k8s/resources/resources.go (hunk 3 above)
Comment (robbrockbank) [reply lib/backend/k8s/resources/resources.go]: Actually, do we need these to be two separate methods?

Can't we just include this processing in the ConvertResourceToKVPair and ConvertKVPairToResource functions.

Comment (robbrockbank) [reply lib/backend/k8s/resources/resources.go]: Hmm, given that this will only be used by the CRD backed resources, I wonder if this should just be part of the CRD conversion methods.  Don't mind it staying here though if you prefer but you should tighten up the method names though to indicate what they're doing.

File: lib/backend/k8s/resources/customresource.go [hunk 12]
```diff
@@ -79,8 +81,17 @@ func (c *customK8sResourceClient) Create(ctx context.Context, kvp *model.KVPair)
 		return nil, K8sErrorToCalico(err, kvp.Key)
 	}
 
+	resOut.GetObjectKind().SetGroupVersionKind(c.k8sResourceTypeMeta.GetObjectKind().GroupVersionKind())
+
+	// Update the return data with the metadata populated by the (Kubernetes) datastore.
+	kvp, err = c.convertResourceToKVPair(resOut)
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/customresource.go]: Do we need this convertResourceToKVPair function - can't we just call directly through to the resource.go functions to convert the resource?  Seems like there's some surplus code here or further down the line.

Comment (robbrockbank) [reply lib/backend/k8s/resources/customresource.go]: or if you are gonna keep it then you should include the other bits above and below in the function.

File: lib/clientv2/globalnetworkpolicy.go [hunk 13]
```diff
@@ -41,8 +42,16 @@ type globalnetworkpolicies struct {
 // representation of the GlobalNetworkPolicy, and an error, if there is any.
 func (r globalnetworkpolicies) Create(ctx context.Context, res *apiv2.GlobalNetworkPolicy, opts options.SetOptions) (*apiv2.GlobalNetworkPolicy, error) {
 	defaultPolicyTypesField(&res.Spec)
+
+	// Properly prefix the name
```
Comment (robbrockbank) [inline lib/clientv2/globalnetworkpolicy.go]: I think we should have converter methods in both directions.

We are disallowing tier names with a kip prefix - so you can use the same check in both directions to add or remove the prefix.

Comment (robbrockbank) [reply lib/clientv2/globalnetworkpolicy.go]: Oh just realized - we only need this for NetworkPolicies not GlobalNetworkPolicies

File: lib/backend/k8s/resources/resources.go (hunk 3 above)
Comment (mgleung) [reply lib/backend/k8s/resources/resources.go]: This logic used to be built into the above methods but then I realized that it also needed to be present in the `customK8sResourceClient` if we wanted to ensure that the `resource` input is not modified (since ideally functions should not leave side effects on the inputs). If we're fine with the input data being modified, then I can move the logic back into the above methods and fix the tests. I was really only made aware of this from the new k8s tests.

File: lib/backend/k8s/resources/resources.go (hunk 5 above)
Comment (mgleung) [reply lib/backend/k8s/resources/resources.go]: I tried using `metav1.Time{}` but it gave me an empty Time object instead of the zero value for a Time object.  Ideally I'm hoping to remove the `CreationTimestamp` and other fields as if they didn't exist instead of filling in the values with empty ones. I did a poor job naming the variables there though so I can fix that.

File: lib/backend/k8s/resources/customresource.go (hunk 6 above)
Comment (mgleung) [reply lib/backend/k8s/resources/customresource.go]: I have it deferred in these functions because I noticed that the input data to the functions was being modified. Since the `Object` interface we're dealing with doesn't have a `DeepCopy` method defined on it, it makes it difficult to modify the contents of the metadata without modifying the input data. I thought it would be cleaner to not let input data to functions modify the input data but if we're always expecting these functions and their outputs to be used in a controlled manner, then I don't think it matters.

File: lib/backend/k8s/resources/resources.go (hunk 10 above)
Comment (mgleung) [reply lib/backend/k8s/resources/resources.go]: I thought about using the `DeepCopyInto` method to create the metadata, but the interface definitions of the `Resource` do not have any methods to replace the `ObjectMeta` struct do they? The issue is I think we need to use the resources for these since the concrete value of each resource is different depending on the type.

File: lib/clientv2/globalnetworkpolicy.go (hunk 11 above)
Comment (mgleung) [reply lib/clientv2/globalnetworkpolicy.go]: Since we are internally prefixing our name data, this prevents the external output from reflecting the internal prefixing that we are adding.

File: lib/backend/k8s/resources/customresource.go (hunk 12 above)
Comment (mgleung) [reply lib/backend/k8s/resources/customresource.go]: We could put this function inline, but since it is called multiple times throughout the other methods, I thought it would be cleaner to keep it all together. There's no real reason that this specifically exists other than to prevent lines of code from being rewritten.

File: lib/backend/k8s/resources/resources.go (hunk 10 above)
Comment (robbrockbank) [reply lib/backend/k8s/resources/resources.go]: No you can't replace the Metadata unfortunately - but you can deep copy into it, and the GetObjectMetadata() returns you the pointer to the ObjectMeta.

File: lib/backend/k8s/resources/customresource.go (hunk 6 above)
Comment (robbrockbank) [reply lib/backend/k8s/resources/customresource.go]: I have a couple of concerns.  Firstly doing this in a defer feels odd to me because it makes the flow of the code hard to follow.  It's clearer if you Modify the ObjectMeta, send the request, Update the ObjectMeta from the response data.  So really the use of defer is more that it's difficult to follow the flow here.  Is your concern that in the error case we've modified the resource that was given to us?  I thought DeepCopy was on the interface now?

File: lib/clientv2/globalnetworkpolicy.go (hunk 11 above)
Comment (robbrockbank) [reply lib/clientv2/globalnetworkpolicy.go]: Actually I see it's because you needed a function to wrap this up for a defer.

File: lib/backend/k8s/resources/customresource.go (hunk 12 above)
Comment (robbrockbank) [reply lib/backend/k8s/resources/customresource.go]: Yup - that makes sense.  I was more worried that we seem to have a bit more redirection than I think we need - that was all.  Because the convertResource then calls out to another convert method to do the heavy lifting

File: lib/backend/k8s/resources/resources.go (hunk 4 above)
Comment (mgleung) [reply lib/backend/k8s/resources/resources.go]: Thought I had to cast it to a concrete type instead of just having it marshal from an interface. Just tested it, marshaling from the interface works.  I'll fix that.

File: lib/backend/k8s/resources/resources.go (hunk 5 above)
Comment (mgleung) [reply lib/backend/k8s/resources/resources.go]: Just retested it. Must've gotten it confused with the slices. `metav1.Time{}` is the empty/zero value.  I'll make that change as well.

File: lib/backend/k8s/resources/resources.go (hunk 10 above)
Comment (mgleung) [reply lib/backend/k8s/resources/resources.go]: Good point, I'll fix that.

File: lib/backend/k8s/resources/customresource.go (hunk 6 above)
Comment (mgleung) [reply lib/backend/k8s/resources/customresource.go]: Yup, my concern was that in the error case (or any case without that function running, I had it in a `defer` just to make sure it was kicked off) we are modifying the input data. I see, I got the `runtime.Object` interface mixed up with the `metav1.Object` interface. I'll change this back.

File: lib/backend/k8s/resources/customresource.go [hunk 14]
```diff
@@ -347,16 +363,25 @@ func (c *customK8sResourceClient) convertResourceToKVPair(r Resource) (*model.KV
...
+	}
+
+	kvp.Value = r
 	return kvp, nil
 }
 
 func (c *customK8sResourceClient) convertKVPairToResource(kvp *model.KVPair) (Resource, error) {
 	resource := kvp.Value.(Resource)
+	// TODO: Figure out if Resource version is suposed to be overwritten by K8s or not.
 	resource.GetObjectMeta().SetResourceVersion(kvp.Revision)
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/customresource.go]: We do need to pass down the resource version into the resource definition that we pass to k8s.  This is used to perform "atomic" operations.  The K8s response will have an updated resource version.

File: lib/backend/k8s/resources/resources.go [hunk 15]
```diff
@@ -100,3 +103,81 @@ func SetCalicoMetadataFromK8sAnnotations(calicoRes Resource, k8sRes Resource) {
 		}
 	}
 }
+
+// Store Calico Metadata in the in the k8s resource annotations for CRD backed resources.
+// This should store all Metadata except for those stored in Annotations and Labels and
+// store them in annotations.
+func StoreCalicoMetadataInK8sAnnotations(resIn Resource) (Resource, error) {
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/resources.go]: Not sure about these function names.  Maybe these should just be ConvertCalicoResourceToK8sResource and ConvertK8sResourceToCalicoResource - obviously we still need to document what that conversion entails.

File: lib/backend/k8s/resources/customresource.go (hunk 14 above)
Comment (robbrockbank) [reply lib/backend/k8s/resources/customresource.go]: The ResourceVersion should be extracted from the response and included in the KVPair that is returned.

Comment (mgleung) [reply lib/backend/k8s/resources/customresource.go]: Sounds good. Wasn't sure if I needed to store the version in the annotations as well, but from your answer, it seems we don't need to.

File: lib/backend/k8s/resources/resources.go [hunk 16]
```diff
@@ -100,3 +103,81 @@ func SetCalicoMetadataFromK8sAnnotations(calicoRes Resource, k8sRes Resource) {
...
+
+	return resOut, nil
+}
+
+// Retrieve all of the Calico Metadata from the k8s resource annotations for CRD backed
+// resources. This should remove the relevant Calico Metadata annotation when it has finished.
+func GetCalicoMetadataFromK8sAnnotations(res Resource) error {
+	rom := res.GetObjectMeta()
+	annotations := rom.GetAnnotations()
+	if len(annotations) == 0 {
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/resources.go]: I think we need to be more forgiving about the annotation not being present.  For example, our k8s manifest may want to include some of the CRDs for pre-population.  In that case we should return the labels/annotations/uuid/name/namespace from the resource but nothing else.

Comment (robbrockbank) [reply lib/backend/k8s/resources/resources.go]: When we write data through libcalico-go we do want to persist the values that are given to us - so as not to potentially confuse the AAPIS -  but reading stuff is less important because that will include data that has already been configured and the AAPIS will learn about that through its watch.

File: lib/backend/k8s/resources/resources.go [hunk 17]
```diff
@@ -100,3 +103,81 @@ func SetCalicoMetadataFromK8sAnnotations(calicoRes Resource, k8sRes Resource) {
...
+
+// Store Calico Metadata in the in the k8s resource annotations for CRD backed resources.
+// This should store all Metadata except for those stored in Annotations and Labels and
+// store them in annotations.
+func StoreCalicoMetadataInK8sAnnotations(resIn Resource) (Resource, error) {
+	rom := resIn.GetObjectMeta()
+
+	// Make sure to remove data that is passed to Kubernetes so it is not duplicated in
+	// the annotations.
+	romCopy := &metav1.ObjectMeta{}
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/resources.go]: Whilst this feels like a little more cumbersome - I'm much happier about that way this is being done now - it feels much more maintainable.  Good stuff.

File: lib/clientv2/globalnetworkpolicy.go [hunk 18]
```diff
@@ -16,6 +16,7 @@ package clientv2
```
Comment (robbrockbank) [inline lib/clientv2/globalnetworkpolicy.go]: GlobalNetworkPolicies don't need any of this special name handling since we don't have any k8s based policies show up here.

Comment (robbrockbank) [reply lib/clientv2/globalnetworkpolicy.go]: Apologies - I thought I'd written that in one of my comments and it looks like i did not.

Comment (robbrockbank) [reply lib/clientv2/globalnetworkpolicy.go]: Oh actually, yes I did.

Comment (mgleung) [reply lib/clientv2/globalnetworkpolicy.go]: Oh oops missed that one. My bad.

File: lib/clientv2/networkpolicy.go [hunk 19]
```diff
@@ -41,8 +42,16 @@ type networkPolicies struct {
 // representation of the NetworkPolicy, and an error, if there is any.
 func (r networkPolicies) Create(ctx context.Context, res *apiv2.NetworkPolicy, opts options.SetOptions) (*apiv2.NetworkPolicy, error) {
 	defaultPolicyTypesField(&res.Spec)
+
+	// Properly prefix the name
+	name := res.GetObjectMeta().GetName()
+	defer resetPolicyName(res, name)
```
Comment (robbrockbank) [inline lib/clientv2/networkpolicy.go]: You don't need this defer anymore.

I think things would feel a lot more symmetric here if you had a convertPolicyNameForStorage and convertPolicyNameFromStorage  or something similar.   You've got one method in one direction and on in another.

That way I think this code would look something like:

`res.GetObjectMeta().SetName(convertPolicyNameForStorage(res.GetObjectMeta().GetName()))`

and just before you return the result you do the same with the opposite conversion function.  That feels like it hides less away.

Obviously for the list you'd need to go through each Item and convert those.

File: lib/clientv2/networkpolicy.go [hunk 20]
```diff
@@ -92,3 +124,42 @@ func (r networkPolicies) List(ctx context.Context, opts options.ListOptions) (*a
...
+			res.Spec.Types = []apiv2.PolicyType{apiv2.PolicyTypeEgress}
+		} else {
+			// Policy has both ingress and egress rules, so apply this policy to both
+			// ingress and egress.
+			res.Spec.Types = []apiv2.PolicyType{apiv2.PolicyTypeIngress, apiv2.PolicyTypeEgress}
+		}
+	}
+}
+
+func removePrefixFromPolicies(res *apiv2.NetworkPolicyList) {
```
Comment (robbrockbank) [inline lib/clientv2/networkpolicy.go]: I would just inline this function in the List.

File: lib/clientv2/networkpolicy.go (hunk 19 above)
Comment (robbrockbank) [reply lib/clientv2/networkpolicy.go]: I'm not suggesting mine is the right way, but it feels cleaner to me than what we have at the moment since we have a number of different methods used at different times performing subtly different things.

File: lib/clientv2/networkpolicy.go [hunk 21]
```diff
@@ -92,3 +124,42 @@ func (r networkPolicies) List(ctx context.Context, opts options.ListOptions) (*a
 func (r networkPolicies) Watch(ctx context.Context, opts options.ListOptions) (watch.Interface, error) {
 	return r.client.resources.Watch(ctx, opts, apiv2.KindNetworkPolicy)
 }
+
+func (r networkPolicies) defaultTypesField(res *apiv2.NetworkPolicy) {
```
Comment (robbrockbank) [inline lib/clientv2/networkpolicy.go]: Where did this change come from?

Comment (robbrockbank) [reply lib/clientv2/networkpolicy.go]: I think this is redundant now.

File: lib/clientv2/networkpolicy.go (hunk 19 above)
Comment (mgleung) [reply lib/clientv2/networkpolicy.go]: I agree that it would be easier to read with symmetric name conversion functions.  Thinking about it, the only reason I can see keeping the way its written now is if for some reason the return value from the client for some reason does not follow the `<prefix>.<name>` format. How valid of a case would it be that data is injected into etcd (I'm not sure if you can do this to KDD) that would not follow our internal prefixing structure and is named something like `<something>.<random>`?

Comment (mgleung) [reply lib/clientv2/networkpolicy.go]: Actually, that's a silly concern because if we convert correctly, then this wouldn't find the strange name that they might manually inject.

File: lib/clientv2/globalnetworkpolicy.go (hunk 18 above)
Comment (mgleung) [reply lib/clientv2/globalnetworkpolicy.go]: From our conversation about this, I will be leaving the prefixing logic in the GlobalNetworkPolicy as well since we need it for non k8s based policies as well.

Comment (robbrockbank) [reply lib/clientv2/globalnetworkpolicy.go]: Yup - my bad.

File: lib/backend/k8s/resources/customresource_test.go [hunk 22]
```diff
@@ -107,11 +108,25 @@ var _ = Describe("BGP Peer conversion methods", func() {
...
+		// Make sure to clean up the annotations on the resource
+		err = ConvertK8sResourceToCalicoResource(r)
+		Expect(err).NotTo(HaveOccurred())
 	})
 
 	It("should convert between a Kuberenetes resource and the equivalent KVPair", func() {
+		// Make sure it returns the same resource if a resource is given without Calico metadata annotations.
 		kvp, err := client.convertResourceToKVPair(res1)
 		Expect(err).NotTo(HaveOccurred())
+		Expect(kvp.Value.(*apiv2.BGPPeer)).To(BeAssignableToTypeOf(&apiv2.BGPPeer{}))
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/customresource_test.go]: I think you mean this to be `Expect(kvp.Value).To(BeAssignableToTypeOf(&apiv2.BGPPeer{}))`

i.e. without casting the Value - otherwise if it isn't a BGPPeer that cast will panic.

Comment (robbrockbank) [reply lib/backend/k8s/resources/customresource_test.go]: Given you are doing a value comparison below - I don't think this particular Expect adds any value.

File: lib/backend/k8s/resources/resources.go [hunk 23]
```diff
@@ -100,3 +102,82 @@ func SetCalicoMetadataFromK8sAnnotations(calicoRes Resource, k8sRes Resource) {
...
+// resources. This should remove the relevant Calico Metadata annotation when it has finished.
+func ConvertK8sResourceToCalicoResource(res Resource) error {
+	rom := res.GetObjectMeta()
+	annotations := rom.GetAnnotations()
+	if len(annotations) == 0 {
+		// Make no changes if there are no annotations to read Calico Metadata out of.
+		return nil
+	}
+	meta := &metav1.ObjectMeta{}
+	err := json.Unmarshal([]byte(annotations[metadataAnnotation]), meta)
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/resources.go]: you should also check if metadataAnnotation is a valid key in the annotations.

File: lib/clientv2/networkpolicy.go [hunk 24]
```diff
@@ -71,8 +123,19 @@ func (r networkPolicies) Delete(ctx context.Context, namespace, name string, opt
 // Get takes name of the NetworkPolicy, and returns the corresponding NetworkPolicy object,
 // and an error if there is any.
 func (r networkPolicies) Get(ctx context.Context, namespace, name string, opts options.GetOptions) (*apiv2.NetworkPolicy, error) {
-	out, err := r.client.resources.Get(ctx, opts, apiv2.KindNetworkPolicy, namespace, name)
+	// Properly prefix the name
+	prefixedName := name
+	if !strings.HasPrefix(name, "knp.") {
```
Comment (robbrockbank) [inline lib/clientv2/networkpolicy.go]: Why isn't this knp check part of the convertPolicyNameForStorage?

File: lib/clientv2/networkpolicy.go [hunk 25]
```diff
@@ -92,3 +165,7 @@ func (r networkPolicies) List(ctx context.Context, opts options.ListOptions) (*a
 func (r networkPolicies) Watch(ctx context.Context, opts options.ListOptions) (watch.Interface, error) {
 	return r.client.resources.Watch(ctx, opts, apiv2.KindNetworkPolicy)
 }
+
+func resetPolicyName(res *apiv2.NetworkPolicy, originalName string) {
```
Comment (robbrockbank) [inline lib/clientv2/networkpolicy.go]: Don't think this is needed anymore

File: lib/backend/k8s/resources/customresource.go [hunk 26]
```diff
@@ -79,8 +79,17 @@ func (c *customK8sResourceClient) Create(ctx context.Context, kvp *model.KVPair)
 		return nil, K8sErrorToCalico(err, kvp.Key)
 	}
 
+	resOut.GetObjectKind().SetGroupVersionKind(c.k8sResourceTypeMeta.GetObjectKind().GroupVersionKind())
```
Comment (robbrockbank) [inline lib/backend/k8s/resources/customresource.go]: Shouldn't this be part of the conversion?
